spark3/src/test/java/org/apache/iceberg/spark/source/TestFilteredScan.java

```diff
@@ -529,7 +529,7 @@ private Table buildPartitionedTable(String desc, PartitionSpec spec, String udf,
     return Lists.newArrayList(
         record(schema, 0L, parse("2017-12-22T09:20:44.294658+00:00"), "junction"),
         record(schema, 1L, parse("2017-12-22T07:15:34.582910+00:00"), "alligator"),
-        record(schema, 2L, parse("2017-12-22T06:02:09.243857+00:00"), "forrest"),
+        record(schema, 2L, parse("2017-12-22T06:02:09.243857+00:00"), ""),
```

Comment:
I can make a separate test if we feel that would be better. However, this test suite covers a very large number of cases so it seems like a smart place to add it.

Reply:
I also think that it's a good opportunity to start introducing more edge case data into the tests without adding any overhead, and it tests this code path in a myriad of ways and does not affect the original test (though I'll have to align my open PR that currently touches this file, but that's not a problem at all).

Reply:
Also, for reference, I did make several separate tests and they all passed. But it just seems cleaner to me to introduce more common edge cases into the existing tests cases. I'm surprised that the tests that run on random data hasn't caught this before. But the tests in `TestFilteredScan` all go over the code paths that deal with this. And after we merge this, I can add in two small / inexpensive tests in the `NOT_STARTS_WITH` pr that will more explicitly cover this edge case - by partitioning a DF using a filter of `data like '%'` and `data not like '%'`, which previously threw an exception due to the truncation length until this change is added.

Reply:
So I actually found an inexpensive existing place to test `truncateBinary` itself. So the function is not better documented by means of the tests (if we agree that this change on the preconditions is acceptable).

Reply:
I'm fine adding the test here, but I'd rather not change the existing cases. Would adding a new record test this without affecting the existing ones?

Reply:
Yes unfortunately, adding a new record to be returned from this function would mean we'd have to change many of the existing tests, as many of them have a hardcoded assumption on 10 records.

I can add a new test in another (possibly new) file if we'd prefer. But in order to place the test in here to be run against what seems to be one of the more comprehensive test suites for hitting various evaluators in various scenarios (i.e. it tests against tables partitioned by the empty string record, tables not partitioned by the empty string record, unpartitioned tables), we'd have to either change an existing record or update the tests to be aware of 11 records.

I'll take another look to see if there's an existing test suite where a new test can be easily added to cover this scenario without affecting other tests.

Reply:
If I add the following test to this file, it does cover some of the codepaths (albeit with a somewhat funny SQL query). This will still go over the `truncateBinary` function when the input needs to be truncated to a length of zero, although this time it will be because the predicate literal having a length of zero instead of the input data having a row with a length of zero.

But the following test would throw prior to this patch (and it doesn't require us to touch any of the input data for the other test suites).

Alternatively, I could add a new test file entirely.

```java
  @Test
  public void testPartitionedByDataStartsWithEmptyStringFilter() {
    File location = buildPartitionedTable("partitioned_by_data", PARTITION_BY_DATA, "data_ident", "data");

    DataSourceOptions options = new DataSourceOptions(ImmutableMap.of(
        "path", location.toString())
    );

    IcebergSource source = new IcebergSource();
    DataSourceReader reader = source.createReader(options);
    pushFilters(reader, new StringStartsWith("data", ""));

    Assert.assertEquals(10, reader.planInputPartitions().size());
  }
```

Reply:
I added a test that does not change any of the exising test data or add a new file at all, in `TestInclusiveMetricsEvaluator`. https://github.com/apache/iceberg/pull/2081/files#diff-a107d871a1a5d5c4029451db68cf5497ddc9432d90619e263270cf72c3536c23R73

I personally prefer changing the input data on the larger end to end test that covers many portions (`TestFilteredScan`) as they truly cover the situation where the bug was originally reported occurring (and where it is likely to happen based on usages of `truncateBinary`) - filtered scans with empty row data.

But I've added another possible way to test the reported scenario -`TestInclusiveMetricsEvaluator`.

If you'd like me to add a standalone spark test or something, let me know. Or if you prefer the new test, let me know and I can revert the change to the input records of the other test suite.

Thanks for your review @rdblue 🙂